src/kv/value/internal/cast/into_primitive.rs

```diff
@@ -0,0 +1,109 @@
+// NOTE: With specialization we could potentially avoid this call using a blanket
```

Comment:
Specialization is possible because we use `impl Trait` instead of `dyn Trait` for inputs. This approach would still be necessary if we switched to `dyn Trait`. The trade-off would be that with `dyn Trait` we could generate less code in calling libraries at the cost of still requiring this runtime check.

Reply:
cc @hawkw sorry for the random ping but thought you might find this interesting for `tracing` 🙂

This is using a static set of type ids sorted at build time and binary searches them for interesting types given a generic `T: Debug + ‘static` input and treats the input as that concrete type. The runtime cost per argument is small (I measured ~9ns on my machine for a hit and ~5ns for a miss) but is not negligible. It could also use specialisation to remove that runtime cost, which I think might land as `min_specialization` sometime over the next year. Maybe you could consider it for something like the `#[instrument]` macro, to keep the nice compatible `Debug` bound, but also retain numbers and booleans as structured values?

Reply:
This is fascinating, thanks for pinging me!

Reply:
So...

If https://github.com/rust-lang/rust/pull/72437 and https://github.com/rust-lang/rust/pull/72488 are both merged then you’d be able to do this at compile time without specialization :)

There’s an example in the `const_type_id` PR.